### PR TITLE
Add Brew distance algorithm with tests

### PR DESCRIPTION
**Pull request by hernanmd**

### CODE REVIEW

File: src/AI-EditDistances/AIBrewDistance.class.st [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+{ #category : #accessing }
+AIBrewDistance >> cost [
+	"Answer an <Array> with the cost of edit operations as: 
+		1st element: Match
+		2nd element: Insert
+		3rd element: Deletion
+		4th element: Substitution "
+
+	^ cost
+		ifNil: [	cost := #( 0 1 1 1 ) ]
```
Comment (jordanmontt) [inline src/AI-EditDistances/AIBrewDistance.class.st:31]: Could we initialize the cost instance variable in the `initialize` method? something like:

``ìnitialize
    cost := self class defaultCostValue.
```
Because if we use this lazy initialization we are forced to use the accessor all the time.

File: src/AI-EditDistances/AIBrewDistance.class.st [hunk 2]
```diff
@@ -0,0 +1,78 @@
+"
+Given two strings, obrain the number of edits required to make the first the same as the second. It use used typically when weighing multiple potentially similar strings in a data set.
```
Comment (jordanmontt) [inline src/AI-EditDistances/AIBrewDistance.class.st:2]: Little typo. obrain -> obtain

It use used typically -> It is used typically

**Review by jordanmontt — CHANGES_REQUESTED** [review]
The code looks good!

I have an observation: actually the Brew algorithm does not exists. It is actually the [Wagner-Fischer algorithm](https://en.wikipedia.org/wiki/Wagner%E2%80%93Fischer_algorithm). The confusion cames from an [article of Chris Brew](https://archive.is/20140611111436/www.ling.ohio-state.edu//%7Ecbrew/795M/string-distance.html#selection-17.172-17.260) when we uses that algorithm with some minor modifications. Here there are some explanations [link](https://github.com/dhgutteridge/Brew-Distance/issues/3). So I would suggest to add this information in the Class comment.

Also this algorithm can also return, optionally, the list of operations (“MATCH”, “DEL”, “INS”, “SUBST) to be applied to make the strings equal. Your implementation already has that information. Could you add  the operations to an array and add an accessor?

**Comment by hernanmd** [comment]
> 

I read the issue thread. I'm not sure to rename the algorithm name since at this point creating another name for it would add more confusion. Besides, although the Brew edit distance and Wagner-Fischer algorithm are both used to compute the edit distance between two strings, there are some differences between them:

- Cost of operations: In the Wagner-Fischer algorithm, the cost of insertion and deletion operations is the same, while in the Brew edit distance, you can have separate costs for deletion (DEL) and insertion (INS) operations. This allows you to differentiate between the two operations and assign different weights to them if needed.
- Substitution operation: The Brew edit distance algorithm includes a separate substitution (SUBST) operation, making it distinct from the match (MATCH) operation. This allows you to assign a different cost to substitutions compared to matches. In the Wagner-Fischer algorithm, a substitution is treated as a combination of deletion and insertion.
- Edit operation description: The Brew edit distance algorithm provides the description of the edits needed to transform the first string into the second one. This feature allows you to understand the specific edit operations that were performed to achieve the final edit distance. The Wagner-Fischer algorithm does not provide this information by default, although it can be modified to track the edit operations as well.

In summary, the Brew edit distance offers more flexibility in terms of operation costs and provides a description of the edit operations, while the Wagner-Fischer algorithm is a more straightforward approach to calculating the edit distance between two strings.

I could add this information in the class comment.

**Review by jecisc — APPROVED** [review]
This seems good to me.

Are you ready to get this merged hernan?

**Comment by hernanmd** [comment]
> This seems good to me.
> 
> Are you ready to get this merged hernan?

Yep, you can merge it as requested changes were pushed.
